File: 75 Must Do Leetcode/3. Longest Substring Without Repeating Characters/longest_substring.rb
def length_of_longest_substring(s)
    r = 0
    l = 0
    max_len = 0
    chars = Set.new
    s.each_char do |c|
        if !chars.include? c
            max_len = [max_len, r - l + 1].max
        else 
            while chars.include? c
                chars.delete(s[l])
                l += 1
            end
        end
        r += 1
        chars.add(c)
    end
    return max_len
end

print length_of_longest_substring("abcabcbb")
puts
print length_of_longest_substring("bbbbb")
puts
print length_of_longest_substring("pwwkew")

# 23/Nov/2021
# Runtime: 124 ms, faster than 56.79% of Ruby online submissions for Longest Substring Without Repeating Characters.
# Memory Usage: 210 MB, less than 96.54% of Ruby online submissions for Longest Substring Without Repeating Characters.

File: 75 Must Do Leetcode/3. Longest Substring Without Repeating Characters/longest_substring.py
def lengthOfLongestSubstring(s: str) -> int:
    chars,l,res = set(),0,0
    
    for r in range(len(s)):
        while s[r] in chars:
            chars.remove(s[l])
            l+=1
        chars.add(s[r])
        res = max(res, r-l+1)
        
    return res

print(lengthOfLongestSubstring("abcabcbb"))
print(lengthOfLongestSubstring("bbbbb"))
print(lengthOfLongestSubstring("pwwkew"))

# 09/Nov/2021
# Runtime: 61 ms, faster than 70.09% of Python3 online submissions for Longest Substring Without Repeating Characters.
# Memory Usage: 14.4 MB, less than 25.47% of Python3 online submissions for Longest Substring Without Repeating Characters.